// @lukso/lsp-factory.js/build/main/src/lib/helpers/config.helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GAS_PRICE = exports.DEFAULT_CONTRACT_VERSION = exports.SET_DATA_PERMISSION = exports.ALL_PERMISSIONS = exports.PREFIX_PERMISSIONS = exports.ADDRESS_PERMISSIONS_ARRAY_KEY = exports.NULL_ADDRESS = exports.LSP3_UP_KEYS = exports.defaultUploadOptions = void 0;
const keccak256_1 = require("@ethersproject/keccak256");
const strings_1 = require("@ethersproject/strings");
const defaultIpfsClientOptions = {
    host: 'api.ipfs.lukso.network',
    port: 443,
    protocol: 'https',
};
exports.defaultUploadOptions = {
    ipfsClientOptions: defaultIpfsClientOptions,
};
exports.LSP3_UP_KEYS = {
    UNIVERSAL_RECEIVER_DELEGATE_KEY: (0, keccak256_1.keccak256)((0, strings_1.toUtf8Bytes)('LSP1UniversalReceiverDelegate')),
    LSP3_PROFILE: (0, keccak256_1.keccak256)((0, strings_1.toUtf8Bytes)('LSP3Profile')),
};
exports.NULL_ADDRESS = '0x0000000000000000000000000000000000000000';
exports.ADDRESS_PERMISSIONS_ARRAY_KEY = '0xdf30dba06db6a30e65354d9a64c609861f089545ca58c6b4dbe31a5f338cb0e3';
exports.PREFIX_PERMISSIONS = '0x4b80742d0000000082ac0000';
exports.ALL_PERMISSIONS = '0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff';
exports.SET_DATA_PERMISSION = '0x0000000000000000000000000000000000000000000000000000000000000008';
exports.DEFAULT_CONTRACT_VERSION = '0.0.1';
exports.GAS_PRICE = 10000000000;
