// @lukso/lsp-factory.js/build/main/src/lib/helpers/deployment.helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBaseContractAddresses = exports.getProxyByteCode = exports.deployProxyContract = exports.deployBaseContract = exports.deployContract = exports.initialize = exports.waitForReceipt = void 0;
const ethers_1 = require("ethers");
const operators_1 = require("rxjs/operators");
const deployment_events_1 = require("../interfaces/deployment-events");
const config_helper_1 = require("./config.helper");
/**
 *
 *
 * @export
 * @template T
 * @param {*} deploymentEvent$
 * @return {*}  {Observable<T>}
 */
function waitForReceipt(deploymentEvent$) {
    return deploymentEvent$.pipe((0, operators_1.switchMap)(async (deploymentEvent) => {
        let status, functionName;
        status = deployment_events_1.DeploymentStatus.COMPLETE;
        switch (deploymentEvent.type) {
            case deployment_events_1.DeploymentType.PROXY:
                functionName = deploymentEvent.functionName;
                status = functionName ? deployment_events_1.DeploymentStatus.COMPLETE : deployment_events_1.DeploymentStatus.PENDING;
                break;
            case deployment_events_1.DeploymentType.TRANSACTION:
                functionName = deploymentEvent.functionName;
                break;
        }
        const receipt = await deploymentEvent.transaction.wait();
        return Object.assign(Object.assign({ type: deploymentEvent.type, contractName: deploymentEvent.contractName }, (functionName && { functionName })), { status,
            receipt });
    }), (0, operators_1.catchError)((error) => {
        const message = 'Error when waiting for the transaction receipt: ' + error.message;
        throw new Error(message);
    }));
}
exports.waitForReceipt = waitForReceipt;
function initialize(deploymentEvent$, factory, initArguments) {
    const initialize$ = deploymentEvent$.pipe((0, operators_1.takeLast)(1), (0, operators_1.switchMap)(async (result) => {
        const contract = await factory.attach(result.receipt.contractAddress);
        const initializeParams = await initArguments(result);
        const transaction = await contract.initialize(...initializeParams, {
            gasLimit: 3000000,
            gasPrice: config_helper_1.GAS_PRICE,
        });
        return {
            type: result.type,
            contractName: result.contractName,
            functionName: 'initialize',
            status: result.status,
            transaction,
        };
    }), (0, operators_1.shareReplay)());
    return initialize$;
}
exports.initialize = initialize;
/**
 * TODO: docs
 */
async function deployContract(deployContractFunction, name) {
    try {
        const contract = await deployContractFunction();
        return {
            type: deployment_events_1.DeploymentType.CONTRACT,
            status: deployment_events_1.DeploymentStatus.PENDING,
            contractName: name,
            transaction: contract.deployTransaction,
        };
    }
    catch (error) {
        console.error(`Error when deploying ${name}`, error);
        throw error;
    }
}
exports.deployContract = deployContract;
async function deployBaseContract(deployContractFunction, name) {
    try {
        const contract = await deployContractFunction();
        return {
            type: deployment_events_1.DeploymentType.BASE_CONTRACT,
            status: deployment_events_1.DeploymentStatus.PENDING,
            contractName: name,
            transaction: contract.deployTransaction,
        };
    }
    catch (error) {
        console.error(`Error when deploying ${name}`, error);
        throw error;
    }
}
exports.deployBaseContract = deployBaseContract;
async function deployProxyContract(abi, deployContractFunction, name, signer) {
    try {
        const contract = await deployContractFunction();
        const factory = new ethers_1.ContractFactory(abi, getProxyByteCode(contract.address), signer);
        const deployedProxy = await factory.deploy();
        const transaction = deployedProxy.deployTransaction;
        return {
            type: deployment_events_1.DeploymentType.PROXY,
            contractName: name,
            status: deployment_events_1.DeploymentStatus.PENDING,
            transaction,
        };
    }
    catch (error) {
        console.error(`Error when deploying ${name}`, error);
        throw error;
    }
}
exports.deployProxyContract = deployProxyContract;
/**
 * Produces the bytecode needed to deploy a minimal proxy contract
 * https://eips.ethereum.org/EIPS/eip-1167
 *
 * @export
 * @param {string} address
 * @return {string}
 */
function getProxyByteCode(address) {
    // prettier-ignore
    return `0x3d602d80600a3d3981f3363d3d373d3d3d363d73${address.substr(2)}5af43d82803e903d91602b57fd5bf3`;
}
exports.getProxyByteCode = getProxyByteCode;
function getBaseContractAddresses(contractDeploymentOptions) {
    return contractDeploymentOptions.libAddresses;
}
exports.getBaseContractAddresses = getBaseContractAddresses;
