// @lukso/lsp-factory.js/build/main/src/lib/helpers/uploader.helper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareImageForLSP3 = exports.ipfsUpload = exports.imageUpload = exports.sizes = void 0;
const browser_image_compression_1 = __importDefault(require("browser-image-compression"));
const utils_1 = require("ethers/lib/utils");
const ipfs_http_client_1 = require("ipfs-http-client");
exports.sizes = [1800, 1024, 640, 320, 180];
async function imageUpload(givenFile, uploadOptions) {
    var _a;
    const isImage = ((_a = givenFile.type) === null || _a === void 0 ? void 0 : _a.substr(0, 6)) === 'image/';
    if (!isImage) {
        throw new Error(`File provided is of type "${givenFile.type}".`);
    }
    return Promise.all(exports.sizes.map(async (size) => {
        const img = await (0, browser_image_compression_1.default)(givenFile, {
            maxWidthOrHeight: size,
            useWebWorker: true,
        });
        const imgBuffer = new Uint8Array(await img.arrayBuffer());
        const loadedImg = await browser_image_compression_1.default.drawFileInCanvas(img);
        const uploadResponse = await ipfsUpload(img, uploadOptions.ipfsClientOptions);
        return {
            width: loadedImg[0].width,
            height: loadedImg[0].height,
            hashFunction: 'keccak256(bytes)',
            hash: (0, utils_1.keccak256)(imgBuffer),
            url: 'ipfs://' + uploadResponse.cid.toString(),
        };
    }));
}
exports.imageUpload = imageUpload;
async function ipfsUpload(file, options) {
    const ipfs = (0, ipfs_http_client_1.create)(options);
    return await ipfs.add(file, {
        pin: true,
    });
}
exports.ipfsUpload = ipfsUpload;
async function prepareImageForLSP3(uploadOptions, image) {
    let lsp3Image;
    if (Array.isArray(image)) {
        lsp3Image = image !== null && image !== void 0 ? image : null;
    }
    else if (image) {
        lsp3Image = await imageUpload(image, uploadOptions);
    }
    return lsp3Image;
}
exports.prepareImageForLSP3 = prepareImageForLSP3;
