// @erc725/erc725.js/build/main/lib/provider-wrapper-utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.constructJSONRPC = exports.decodeResult = void 0;
const abi = __importStar(require("web3-eth-abi"));
const constants_1 = require("./constants");
let idCount = 0;
// @ts-ignore
const web3abiDecoder = abi.default;
function decodeResult(method, result) {
    const rpcResult = result.result;
    if (rpcResult === '0x') {
        return null;
    }
    const decodedData = web3abiDecoder.decodeParameter(constants_1.METHODS[method].returnEncoding, rpcResult);
    if (Array.isArray(decodedData) &&
        decodedData.length === 1 &&
        decodedData[0] === '0x') {
        return [null];
    }
    return decodedData;
}
exports.decodeResult = decodeResult;
function constructJSONRPC(address, method, methodParam) {
    const data = methodParam
        ? constants_1.METHODS[method].sig + methodParam.replace('0x', '')
        : constants_1.METHODS[method].sig;
    idCount += 1;
    return {
        jsonrpc: '2.0',
        method: 'eth_call',
        params: [
            {
                to: address,
                gas: constants_1.METHODS[method].gas,
                gasPrice: constants_1.METHODS[method].gasPrice,
                value: constants_1.METHODS[method].value,
                data,
            },
        ],
        id: idCount,
    };
}
exports.constructJSONRPC = constructJSONRPC;
