// @lukso/lsp-factory.js/build/module/src/lib/helpers/deployment.helper.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { ContractFactory } from 'ethers';
import { catchError, shareReplay, switchMap, takeLast } from 'rxjs/operators';
import { DeploymentStatus, DeploymentType, } from '../interfaces/deployment-events';
import { GAS_PRICE } from './config.helper';
/**
 *
 *
 * @export
 * @template T
 * @param {*} deploymentEvent$
 * @return {*}  {Observable<T>}
 */
export function waitForReceipt(deploymentEvent$) {
    return deploymentEvent$.pipe(switchMap((deploymentEvent) => __awaiter(this, void 0, void 0, function* () {
        let status, functionName;
        status = DeploymentStatus.COMPLETE;
        switch (deploymentEvent.type) {
            case DeploymentType.PROXY:
                functionName = deploymentEvent.functionName;
                status = functionName ? DeploymentStatus.COMPLETE : DeploymentStatus.PENDING;
                break;
            case DeploymentType.TRANSACTION:
                functionName = deploymentEvent.functionName;
                break;
        }
        const receipt = yield deploymentEvent.transaction.wait();
        return Object.assign(Object.assign({ type: deploymentEvent.type, contractName: deploymentEvent.contractName }, (functionName && { functionName })), { status,
            receipt });
    })), catchError((error) => {
        const message = 'Error when waiting for the transaction receipt: ' + error.message;
        throw new Error(message);
    }));
}
export function initialize(deploymentEvent$, factory, initArguments) {
    const initialize$ = deploymentEvent$.pipe(takeLast(1), switchMap((result) => __awaiter(this, void 0, void 0, function* () {
        const contract = yield factory.attach(result.receipt.contractAddress);
        const initializeParams = yield initArguments(result);
        const transaction = yield contract.initialize(...initializeParams, {
            gasLimit: 3000000,
            gasPrice: GAS_PRICE,
        });
        return {
            type: result.type,
            contractName: result.contractName,
            functionName: 'initialize',
            status: result.status,
            transaction,
        };
    })), shareReplay());
    return initialize$;
}
/**
 * TODO: docs
 */
export function deployContract(deployContractFunction, name) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const contract = yield deployContractFunction();
            return {
                type: DeploymentType.CONTRACT,
                status: DeploymentStatus.PENDING,
                contractName: name,
                transaction: contract.deployTransaction,
            };
        }
        catch (error) {
            console.error(`Error when deploying ${name}`, error);
            throw error;
        }
    });
}
export function deployBaseContract(deployContractFunction, name) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const contract = yield deployContractFunction();
            return {
                type: DeploymentType.BASE_CONTRACT,
                status: DeploymentStatus.PENDING,
                contractName: name,
                transaction: contract.deployTransaction,
            };
        }
        catch (error) {
            console.error(`Error when deploying ${name}`, error);
            throw error;
        }
    });
}
export function deployProxyContract(abi, deployContractFunction, name, signer) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const contract = yield deployContractFunction();
            const factory = new ContractFactory(abi, getProxyByteCode(contract.address), signer);
            const deployedProxy = yield factory.deploy();
            const transaction = deployedProxy.deployTransaction;
            return {
                type: DeploymentType.PROXY,
                contractName: name,
                status: DeploymentStatus.PENDING,
                transaction,
            };
        }
        catch (error) {
            console.error(`Error when deploying ${name}`, error);
            throw error;
        }
    });
}
/**
 * Produces the bytecode needed to deploy a minimal proxy contract
 * https://eips.ethereum.org/EIPS/eip-1167
 *
 * @export
 * @param {string} address
 * @return {string}
 */
export function getProxyByteCode(address) {
    // prettier-ignore
    return `0x3d602d80600a3d3981f3363d3d373d3d3d363d73${address.substr(2)}5af43d82803e903d91602b57fd5bf3`;
}
export function getBaseContractAddresses(contractDeploymentOptions) {
    return contractDeploymentOptions.libAddresses;
}
