// @lukso/lsp-factory.js/build/module/src/lib/helpers/uploader.helper.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import imageCompression from 'browser-image-compression';
import { keccak256 } from 'ethers/lib/utils';
import { create } from 'ipfs-http-client';
export const sizes = [1800, 1024, 640, 320, 180];
export function imageUpload(givenFile, uploadOptions) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
        const isImage = ((_a = givenFile.type) === null || _a === void 0 ? void 0 : _a.substr(0, 6)) === 'image/';
        if (!isImage) {
            throw new Error(`File provided is of type "${givenFile.type}".`);
        }
        return Promise.all(sizes.map((size) => __awaiter(this, void 0, void 0, function* () {
            const img = yield imageCompression(givenFile, {
                maxWidthOrHeight: size,
                useWebWorker: true,
            });
            const imgBuffer = new Uint8Array(yield img.arrayBuffer());
            const loadedImg = yield imageCompression.drawFileInCanvas(img);
            const uploadResponse = yield ipfsUpload(img, uploadOptions.ipfsClientOptions);
            return {
                width: loadedImg[0].width,
                height: loadedImg[0].height,
                hashFunction: 'keccak256(bytes)',
                hash: keccak256(imgBuffer),
                url: 'ipfs://' + uploadResponse.cid.toString(),
            };
        })));
    });
}
export function ipfsUpload(file, options) {
    return __awaiter(this, void 0, void 0, function* () {
        const ipfs = create(options);
        return yield ipfs.add(file, {
            pin: true,
        });
    });
}
export function prepareImageForLSP3(uploadOptions, image) {
    return __awaiter(this, void 0, void 0, function* () {
        let lsp3Image;
        if (Array.isArray(image)) {
            lsp3Image = image !== null && image !== void 0 ? image : null;
        }
        else if (image) {
            lsp3Image = yield imageUpload(image, uploadOptions);
        }
        return lsp3Image;
    });
}
