// @erc725/erc725.js/build/module/lib/provider-wrapper-utils.js
import * as abi from 'web3-eth-abi';
import { METHODS } from './constants';
let idCount = 0;
// @ts-ignore
const web3abiDecoder = abi.default;
export function decodeResult(method, result) {
    const rpcResult = result.result;
    if (rpcResult === '0x') {
        return null;
    }
    const decodedData = web3abiDecoder.decodeParameter(METHODS[method].returnEncoding, rpcResult);
    if (Array.isArray(decodedData) &&
        decodedData.length === 1 &&
        decodedData[0] === '0x') {
        return [null];
    }
    return decodedData;
}
export function constructJSONRPC(address, method, methodParam) {
    const data = methodParam
        ? METHODS[method].sig + methodParam.replace('0x', '')
        : METHODS[method].sig;
    idCount += 1;
    return {
        jsonrpc: '2.0',
        method: 'eth_call',
        params: [
            {
                to: address,
                gas: METHODS[method].gas,
                gasPrice: METHODS[method].gasPrice,
                value: METHODS[method].value,
                data,
            },
        ],
        id: idCount,
    };
}
