// @lukso/lsp-factory.js/build/main/src/lib/services/digital-asset.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.lsp8IdentifiableDigitalAssetDeployment$ = exports.lsp7DigitalAssetDeployment$ = void 0;
const rxjs_1 = require("rxjs");
const __1 = require("../../");
const deployment_helper_1 = require("../helpers/deployment.helper");
const digital_asset_deployment_1 = require("../interfaces/digital-asset-deployment");
// LSP7
function lsp7DigitalAssetDeployment$(signer, digitalAssetDeploymentOptions, baseContractAddress) {
    const lsp7Deployment$ = (0, rxjs_1.from)(deployLSP7DigitalAsset(signer, digitalAssetDeploymentOptions, baseContractAddress)).pipe((0, rxjs_1.shareReplay)());
    const lsp7DeploymentReceipt$ = (0, deployment_helper_1.waitForReceipt)(lsp7Deployment$).pipe((0, rxjs_1.shareReplay)());
    const lsp7Initialize$ = baseContractAddress
        ? initializeLSP7Proxy(signer, lsp7DeploymentReceipt$, digitalAssetDeploymentOptions)
        : rxjs_1.EMPTY;
    const lsp7InitializeReceipt$ = (0, deployment_helper_1.waitForReceipt)(lsp7Initialize$).pipe((0, rxjs_1.shareReplay)());
    return (0, rxjs_1.concat)(lsp7Deployment$, lsp7DeploymentReceipt$, lsp7Initialize$, lsp7InitializeReceipt$);
}
exports.lsp7DigitalAssetDeployment$ = lsp7DigitalAssetDeployment$;
async function deployLSP7DigitalAsset(signer, digitalAssetDeploymentOptions, baseContractAddress) {
    const deploymentFunction = async () => {
        return baseContractAddress
            ? new __1.LSP7DigitalAssetInit__factory(signer).attach(baseContractAddress)
            : await new __1.LSP7DigitalAsset__factory(signer).deploy(digitalAssetDeploymentOptions.name, digitalAssetDeploymentOptions.symbol, digitalAssetDeploymentOptions.ownerAddress, digitalAssetDeploymentOptions.isNFT);
    };
    return baseContractAddress
        ? (0, deployment_helper_1.deployProxyContract)(__1.LSP7DigitalAssetInit__factory.abi, deploymentFunction, digital_asset_deployment_1.ContractNames.LSP7_DIGITAL_ASSET, signer)
        : (0, deployment_helper_1.deployContract)(deploymentFunction, digital_asset_deployment_1.ContractNames.LSP7_DIGITAL_ASSET);
}
function initializeLSP7Proxy(signer, digitalAssetDeploymentReceipt$, digitalAssetDeploymentOptions) {
    const { name, symbol, ownerAddress, isNFT } = digitalAssetDeploymentOptions;
    const initialize$ = digitalAssetDeploymentReceipt$.pipe((0, rxjs_1.takeLast)(1), (0, rxjs_1.switchMap)(async (result) => {
        const contract = await new __1.LSP7DigitalAssetInit__factory(signer).attach(result.receipt.contractAddress);
        const transaction = await contract[`initialize(string,string,address,bool)`](name, symbol, ownerAddress, isNFT, {
            gasLimit: 3000000,
        });
        return {
            type: result.type,
            contractName: result.contractName,
            functionName: 'initialize',
            status: result.status,
            transaction,
        };
    }), (0, rxjs_1.shareReplay)());
    return initialize$;
}
// LSP8
function lsp8IdentifiableDigitalAssetDeployment$(signer, digitalAssetDeploymentOptions, baseContractAddress) {
    const lsp8Deployment$ = (0, rxjs_1.from)(deployLSP8IdentifiableDigitalAsset(signer, digitalAssetDeploymentOptions, baseContractAddress)).pipe((0, rxjs_1.shareReplay)());
    const lsp8DeploymentReceipt$ = (0, deployment_helper_1.waitForReceipt)(lsp8Deployment$);
    const lsp8Initialize$ = baseContractAddress
        ? initializeLSP8Proxy(signer, lsp8DeploymentReceipt$, digitalAssetDeploymentOptions)
        : rxjs_1.EMPTY;
    const lsp8InitializeReceipt$ = (0, deployment_helper_1.waitForReceipt)(lsp8Initialize$).pipe((0, rxjs_1.shareReplay)());
    return (0, rxjs_1.concat)(lsp8Deployment$, lsp8DeploymentReceipt$, lsp8Initialize$, lsp8InitializeReceipt$);
}
exports.lsp8IdentifiableDigitalAssetDeployment$ = lsp8IdentifiableDigitalAssetDeployment$;
async function deployLSP8IdentifiableDigitalAsset(signer, digitalAssetDeploymentOptions, baseContractAddress) {
    const deploymentFunction = async () => {
        return baseContractAddress
            ? new __1.LSP8IdentifiableDigitalAssetInit__factory(signer).attach(baseContractAddress)
            : await new __1.LSP8IdentifiableDigitalAsset__factory(signer).deploy(digitalAssetDeploymentOptions.name, digitalAssetDeploymentOptions.symbol, digitalAssetDeploymentOptions.ownerAddress);
    };
    return baseContractAddress
        ? (0, deployment_helper_1.deployProxyContract)(__1.LSP8IdentifiableDigitalAssetInit__factory.abi, deploymentFunction, digital_asset_deployment_1.ContractNames.LSP8_DIGITAL_ASSET, signer)
        : (0, deployment_helper_1.deployContract)(deploymentFunction, digital_asset_deployment_1.ContractNames.LSP8_DIGITAL_ASSET);
}
function initializeLSP8Proxy(signer, digitalAssetDeploymentReceipt$, digitalAssetDeploymentOptions) {
    const { name, symbol, ownerAddress } = digitalAssetDeploymentOptions;
    const initialize$ = digitalAssetDeploymentReceipt$.pipe((0, rxjs_1.takeLast)(1), (0, rxjs_1.switchMap)(async (result) => {
        const contract = await new __1.LSP7DigitalAssetInit__factory(signer).attach(result.receipt.contractAddress);
        const transaction = await contract[`initialize(string,string,address)`](name, symbol, ownerAddress, {
            gasLimit: 3000000,
        });
        return {
            type: result.type,
            contractName: result.contractName,
            functionName: 'initialize',
            status: result.status,
            transaction,
        };
    }), (0, rxjs_1.shareReplay)());
    return initialize$;
}
