// @lukso/lsp-factory.js/build/module/src/lib/services/digital-asset.service.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { concat, EMPTY, from, shareReplay, switchMap, takeLast } from 'rxjs';
import { LSP7DigitalAsset__factory, LSP7DigitalAssetInit__factory, LSP8IdentifiableDigitalAsset__factory, LSP8IdentifiableDigitalAssetInit__factory, } from '../../';
import { deployContract, deployProxyContract, waitForReceipt } from '../helpers/deployment.helper';
import { ContractNames, } from '../interfaces/digital-asset-deployment';
// LSP7
export function lsp7DigitalAssetDeployment$(signer, digitalAssetDeploymentOptions, baseContractAddress) {
    const lsp7Deployment$ = from(deployLSP7DigitalAsset(signer, digitalAssetDeploymentOptions, baseContractAddress)).pipe(shareReplay());
    const lsp7DeploymentReceipt$ = waitForReceipt(lsp7Deployment$).pipe(shareReplay());
    const lsp7Initialize$ = baseContractAddress
        ? initializeLSP7Proxy(signer, lsp7DeploymentReceipt$, digitalAssetDeploymentOptions)
        : EMPTY;
    const lsp7InitializeReceipt$ = waitForReceipt(lsp7Initialize$).pipe(shareReplay());
    return concat(lsp7Deployment$, lsp7DeploymentReceipt$, lsp7Initialize$, lsp7InitializeReceipt$);
}
function deployLSP7DigitalAsset(signer, digitalAssetDeploymentOptions, baseContractAddress) {
    return __awaiter(this, void 0, void 0, function* () {
        const deploymentFunction = () => __awaiter(this, void 0, void 0, function* () {
            return baseContractAddress
                ? new LSP7DigitalAssetInit__factory(signer).attach(baseContractAddress)
                : yield new LSP7DigitalAsset__factory(signer).deploy(digitalAssetDeploymentOptions.name, digitalAssetDeploymentOptions.symbol, digitalAssetDeploymentOptions.ownerAddress, digitalAssetDeploymentOptions.isNFT);
        });
        return baseContractAddress
            ? deployProxyContract(LSP7DigitalAssetInit__factory.abi, deploymentFunction, ContractNames.LSP7_DIGITAL_ASSET, signer)
            : deployContract(deploymentFunction, ContractNames.LSP7_DIGITAL_ASSET);
    });
}
function initializeLSP7Proxy(signer, digitalAssetDeploymentReceipt$, digitalAssetDeploymentOptions) {
    const { name, symbol, ownerAddress, isNFT } = digitalAssetDeploymentOptions;
    const initialize$ = digitalAssetDeploymentReceipt$.pipe(takeLast(1), switchMap((result) => __awaiter(this, void 0, void 0, function* () {
        const contract = yield new LSP7DigitalAssetInit__factory(signer).attach(result.receipt.contractAddress);
        const transaction = yield contract[`initialize(string,string,address,bool)`](name, symbol, ownerAddress, isNFT, {
            gasLimit: 3000000,
        });
        return {
            type: result.type,
            contractName: result.contractName,
            functionName: 'initialize',
            status: result.status,
            transaction,
        };
    })), shareReplay());
    return initialize$;
}
// LSP8
export function lsp8IdentifiableDigitalAssetDeployment$(signer, digitalAssetDeploymentOptions, baseContractAddress) {
    const lsp8Deployment$ = from(deployLSP8IdentifiableDigitalAsset(signer, digitalAssetDeploymentOptions, baseContractAddress)).pipe(shareReplay());
    const lsp8DeploymentReceipt$ = waitForReceipt(lsp8Deployment$);
    const lsp8Initialize$ = baseContractAddress
        ? initializeLSP8Proxy(signer, lsp8DeploymentReceipt$, digitalAssetDeploymentOptions)
        : EMPTY;
    const lsp8InitializeReceipt$ = waitForReceipt(lsp8Initialize$).pipe(shareReplay());
    return concat(lsp8Deployment$, lsp8DeploymentReceipt$, lsp8Initialize$, lsp8InitializeReceipt$);
}
function deployLSP8IdentifiableDigitalAsset(signer, digitalAssetDeploymentOptions, baseContractAddress) {
    return __awaiter(this, void 0, void 0, function* () {
        const deploymentFunction = () => __awaiter(this, void 0, void 0, function* () {
            return baseContractAddress
                ? new LSP8IdentifiableDigitalAssetInit__factory(signer).attach(baseContractAddress)
                : yield new LSP8IdentifiableDigitalAsset__factory(signer).deploy(digitalAssetDeploymentOptions.name, digitalAssetDeploymentOptions.symbol, digitalAssetDeploymentOptions.ownerAddress);
        });
        return baseContractAddress
            ? deployProxyContract(LSP8IdentifiableDigitalAssetInit__factory.abi, deploymentFunction, ContractNames.LSP8_DIGITAL_ASSET, signer)
            : deployContract(deploymentFunction, ContractNames.LSP8_DIGITAL_ASSET);
    });
}
function initializeLSP8Proxy(signer, digitalAssetDeploymentReceipt$, digitalAssetDeploymentOptions) {
    const { name, symbol, ownerAddress } = digitalAssetDeploymentOptions;
    const initialize$ = digitalAssetDeploymentReceipt$.pipe(takeLast(1), switchMap((result) => __awaiter(this, void 0, void 0, function* () {
        const contract = yield new LSP7DigitalAssetInit__factory(signer).attach(result.receipt.contractAddress);
        const transaction = yield contract[`initialize(string,string,address)`](name, symbol, ownerAddress, {
            gasLimit: 3000000,
        });
        return {
            type: result.type,
            contractName: result.contractName,
            functionName: 'initialize',
            status: result.status,
            transaction,
        };
    })), shareReplay());
    return initialize$;
}
