// @erc725/erc725.js/build/main/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HASH_FUNCTIONS = exports.SUPPORTED_HASH_FUNCTIONS_LIST = exports.SUPPORTED_HASH_FUNCTION_HASHES = exports.SUPPORTED_HASH_FUNCTION_STRINGS = exports.METHODS = void 0;
/* eslint-disable @typescript-eslint/ban-types */
const web3_utils_1 = require("web3-utils");
const Method_1 = require("../types/Method");
exports.METHODS = {
    [Method_1.Method.GET_DATA]: {
        sig: '0x54f6127f',
        gas: web3_utils_1.numberToHex(2000000),
        gasPrice: web3_utils_1.numberToHex(100000000),
        value: web3_utils_1.numberToHex(0),
        returnEncoding: Method_1.Encoding.BYTES,
    },
    [Method_1.Method.DATA_COUNT]: {
        sig: '0x5da40c47',
        gas: web3_utils_1.numberToHex(2000000),
        gasPrice: web3_utils_1.numberToHex(100000000),
        value: web3_utils_1.numberToHex(0),
        returnEncoding: Method_1.Encoding.UINT256,
    },
    [Method_1.Method.ALL_DATA]: {
        sig: '0xc559acef',
        gas: web3_utils_1.numberToHex(2000000),
        gasPrice: web3_utils_1.numberToHex(100000000),
        value: web3_utils_1.numberToHex(0),
        returnEncoding: Method_1.Encoding.BYTES32_ARRAY,
    },
    [Method_1.Method.OWNER]: {
        sig: '0x8da5cb5b',
        gas: web3_utils_1.numberToHex(2000000),
        gasPrice: web3_utils_1.numberToHex(100000000),
        value: web3_utils_1.numberToHex(0),
        returnEncoding: Method_1.Encoding.ADDRESS,
    },
};
var SUPPORTED_HASH_FUNCTION_STRINGS;
(function (SUPPORTED_HASH_FUNCTION_STRINGS) {
    SUPPORTED_HASH_FUNCTION_STRINGS["KECCAK256_UTF8"] = "keccak256(utf8)";
    SUPPORTED_HASH_FUNCTION_STRINGS["KECCAK256_BYTES"] = "keccak256(bytes)";
})(SUPPORTED_HASH_FUNCTION_STRINGS = exports.SUPPORTED_HASH_FUNCTION_STRINGS || (exports.SUPPORTED_HASH_FUNCTION_STRINGS = {}));
var SUPPORTED_HASH_FUNCTION_HASHES;
(function (SUPPORTED_HASH_FUNCTION_HASHES) {
    SUPPORTED_HASH_FUNCTION_HASHES["HASH_KECCAK256_UTF8"] = "0x6f357c6a";
    SUPPORTED_HASH_FUNCTION_HASHES["HASH_KECCAK256_BYTES"] = "0x8019f9b1";
})(SUPPORTED_HASH_FUNCTION_HASHES = exports.SUPPORTED_HASH_FUNCTION_HASHES || (exports.SUPPORTED_HASH_FUNCTION_HASHES = {}));
exports.SUPPORTED_HASH_FUNCTIONS_LIST = Object.values(SUPPORTED_HASH_FUNCTION_STRINGS);
function keccak256Utf8(data) {
    return web3_utils_1.keccak256(JSON.stringify(data));
}
const KECCAK256_UTF8 = {
    method: keccak256Utf8,
    name: SUPPORTED_HASH_FUNCTION_STRINGS.KECCAK256_UTF8,
    sig: SUPPORTED_HASH_FUNCTION_HASHES.HASH_KECCAK256_UTF8,
};
const KECCAK256_BYTES = {
    method: web3_utils_1.keccak256,
    name: SUPPORTED_HASH_FUNCTION_STRINGS.KECCAK256_BYTES,
    sig: SUPPORTED_HASH_FUNCTION_HASHES.HASH_KECCAK256_BYTES,
};
exports.HASH_FUNCTIONS = {
    [SUPPORTED_HASH_FUNCTION_STRINGS.KECCAK256_UTF8]: KECCAK256_UTF8,
    [SUPPORTED_HASH_FUNCTION_HASHES.HASH_KECCAK256_UTF8]: KECCAK256_UTF8,
    [SUPPORTED_HASH_FUNCTION_STRINGS.KECCAK256_BYTES]: KECCAK256_BYTES,
    [SUPPORTED_HASH_FUNCTION_HASHES.HASH_KECCAK256_BYTES]: KECCAK256_BYTES,
};
