// @erc725/erc725.js/build/main/providers/ethereumProviderWrapper.js
"use strict";
/*
    This file is part of @erc725/erc725.js.
    @erc725/erc725.js is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
    @erc725/erc725.js is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.
    You should have received a copy of the GNU Lesser General Public License
    along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/
/**
 * @file providers/ethereumProviderWrapper.ts
 * @author Robert McLeod <@robertdavid010>, Fabian Vogelsteller <fabian@lukso.network>
 * @date 2020
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthereumProviderWrapper = void 0;
/*
  This file will handle querying the Ethereum provider api in accordance with
  implementation of smart contract interfaces of ERC725
*/
const abi = __importStar(require("web3-eth-abi"));
const constants_1 = require("../lib/constants");
const Method_1 = require("../types/Method");
// @ts-ignore
const web3Abi = abi.default;
class EthereumProviderWrapper {
    constructor(provider) {
        this.type = "ETHEREUM" /* ETHEREUM */;
        this.provider = provider;
    }
    async getOwner(address) {
        const params = this.constructJSONRPC(address, Method_1.Method.OWNER);
        const result = await this.callContract([params]);
        if (result.error) {
            throw result.error;
        }
        return this.decodeResult(Method_1.Method.OWNER, result);
    }
    async getData(address, keyHash) {
        const result = await this.callContract([
            this.constructJSONRPC(address, Method_1.Method.GET_DATA, keyHash),
        ]);
        return this.decodeResult(Method_1.Method.GET_DATA, result);
    }
    async getAllData(address, keys) {
        const results = [];
        for (let index = 0; index < keys.length; index++) {
            results.push({
                key: keys[index],
                value: await this.getData(address, keys[index]),
            });
        }
        return results;
    }
    // eslint-disable-next-line class-methods-use-this
    constructJSONRPC(address, method, methodParam) {
        const data = methodParam
            ? constants_1.METHODS[method].sig + methodParam.replace('0x', '')
            : constants_1.METHODS[method].sig;
        return {
            to: address,
            gas: constants_1.METHODS[method].gas,
            gasPrice: constants_1.METHODS[method].gasPrice,
            value: constants_1.METHODS[method].value,
            data,
        };
    }
    async callContract(params) {
        return this.provider.request({ method: 'eth_call', params });
    }
    // eslint-disable-next-line class-methods-use-this
    decodeResult(method, result) {
        return result === '0x'
            ? null
            : web3Abi.decodeParameter(constants_1.METHODS[method].returnEncoding, result);
    }
}
exports.EthereumProviderWrapper = EthereumProviderWrapper;
