// @erc725/erc725.js/build/main/providers/web3ProviderWrapper.js
"use strict";
/*
    This file is part of @erc725/erc725.js.
    @erc725/erc725.js is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
    @erc725/erc725.js is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.
    You should have received a copy of the GNU Lesser General Public License
    along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/
/**
 * @file providers/web3ProviderWrapper.ts
 * @author Robert McLeod <@robertdavid010>, Fabian Vogelsteller <fabian@lukso.network>
 * @date 2020
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Web3ProviderWrapper = void 0;
const Method_1 = require("../types/Method");
const provider_wrapper_utils_1 = require("../lib/provider-wrapper-utils");
class Web3ProviderWrapper {
    constructor(provider) {
        this.type = "WEB3" /* WEB3 */;
        this.provider = provider;
    }
    async getOwner(address) {
        const result = await this.callContract(provider_wrapper_utils_1.constructJSONRPC(address, Method_1.Method.OWNER));
        if (result.error) {
            throw result.error;
        }
        return provider_wrapper_utils_1.decodeResult(Method_1.Method.OWNER, result);
    }
    async getData(address, keyHash) {
        return provider_wrapper_utils_1.decodeResult(Method_1.Method.GET_DATA, await this.callContract(provider_wrapper_utils_1.constructJSONRPC(address, Method_1.Method.GET_DATA, keyHash)));
    }
    async getAllData(address, keys) {
        const payload = [];
        for (let index = 0; index < keys.length; index++) {
            payload.push(provider_wrapper_utils_1.constructJSONRPC(address, Method_1.Method.GET_DATA, keys[index]));
        }
        const results = await this.callContract(payload);
        // map results to keys
        const returnValues = [];
        for (let index = 0; index < payload.length; index++) {
            returnValues.push({
                key: keys[index],
                value: provider_wrapper_utils_1.decodeResult(Method_1.Method.GET_DATA, results.find((element) => payload[index].id === element.id)),
            });
        }
        return returnValues;
    }
    async callContract(payload) {
        return new Promise((resolve, reject) => {
            // Send old web3 method with callback to resolve promise
            this.provider.send(payload, (e, r) => {
                if (e) {
                    reject(e);
                }
                else {
                    resolve(r);
                }
            });
        });
    }
}
exports.Web3ProviderWrapper = Web3ProviderWrapper;
