// @lukso/lsp-factory.js/build/module/src/lib/classes/digital-asset.js
import { NonceManager } from '@ethersproject/experimental';
import { lastValueFrom, of, scan } from 'rxjs';
import versions from '../../versions.json';
import { DEFAULT_CONTRACT_VERSION } from '../helpers/config.helper';
import { digitalAssetBaseContractsDeployment$ } from '../services/base-contract.service';
import { lsp7DigitalAssetDeployment$, lsp8IdentifiableDigitalAssetDeployment$, } from '../services/digital-asset.service';
export class DigitalAsset {
    constructor(options) {
        this.options = options;
        this.signer = new NonceManager(options.signer);
    }
    // LSP7
    deployLSP7DigitalAssetReactive(digitalAssetDeploymentOptions, contractDeploymentOptions) {
        var _a, _b, _c;
        const digitalAsset$ = lsp7DigitalAssetDeployment$(this.signer, digitalAssetDeploymentOptions, (_a = contractDeploymentOptions === null || contractDeploymentOptions === void 0 ? void 0 : contractDeploymentOptions.libAddress) !== null && _a !== void 0 ? _a : (_c = (_b = versions[this.options.chainId]) === null || _b === void 0 ? void 0 : _b.baseContracts) === null || _c === void 0 ? void 0 : _c.LSP7DigitalAsset[DEFAULT_CONTRACT_VERSION]);
        return digitalAsset$;
    }
    deployLSP7DigitalAsset(digitalAssetDeploymentOptions, contractDeploymentOptions) {
        const deployments$ = this.deployLSP7DigitalAssetReactive(digitalAssetDeploymentOptions, contractDeploymentOptions).pipe(scan((accumulator, deploymentEvent) => {
            if (deploymentEvent.receipt && deploymentEvent.receipt.contractAddress) {
                accumulator[deploymentEvent.contractName] = {
                    address: deploymentEvent.receipt.contractAddress,
                    receipt: deploymentEvent.receipt,
                };
            }
            return accumulator;
        }, {}));
        return lastValueFrom(deployments$);
    }
    // LSP8
    deployLSP8IdentifiableDigitalAssetReactive(digitalAssetDeploymentOptions, contractDeploymentOptions) {
        var _a, _b, _c;
        const digitalAsset$ = lsp8IdentifiableDigitalAssetDeployment$(this.signer, digitalAssetDeploymentOptions, (_a = contractDeploymentOptions === null || contractDeploymentOptions === void 0 ? void 0 : contractDeploymentOptions.libAddress) !== null && _a !== void 0 ? _a : (_c = (_b = versions[this.options.chainId]) === null || _b === void 0 ? void 0 : _b.baseContracts) === null || _c === void 0 ? void 0 : _c.LSP8IdentifiableDigitalAsset[DEFAULT_CONTRACT_VERSION]);
        return digitalAsset$;
    }
    deployLSP8IdentifiableDigitalAsset(digitalAssetDeploymentOptions, ContractDeploymentOptions) {
        const deployments$ = this.deployLSP8IdentifiableDigitalAssetReactive(digitalAssetDeploymentOptions, ContractDeploymentOptions).pipe(scan((accumulator, deploymentEvent) => {
            if (deploymentEvent.receipt && deploymentEvent.receipt.contractAddress) {
                accumulator[deploymentEvent.contractName] = {
                    address: deploymentEvent.receipt.contractAddress,
                    receipt: deploymentEvent.receipt,
                };
            }
            return accumulator;
        }, {}));
        return lastValueFrom(deployments$);
    }
    deployBaseContracts() {
        const baseContractsToDeploy$ = of([true, true]);
        const baseContracts$ = digitalAssetBaseContractsDeployment$(this.signer, baseContractsToDeploy$);
        const deployments$ = baseContracts$.pipe(scan((accumulator, deploymentEvent) => {
            if (deploymentEvent.receipt && deploymentEvent.receipt.contractAddress) {
                accumulator[deploymentEvent.contractName] = {
                    address: deploymentEvent.receipt.contractAddress,
                    receipt: deploymentEvent.receipt,
                };
            }
            return accumulator;
        }, {}));
        return lastValueFrom(deployments$);
    }
}
