// @lukso/lsp-factory.js/build/module/src/lib/classes/lsp3-universal-profile.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { NonceManager } from '@ethersproject/experimental';
import { concat, forkJoin, lastValueFrom, merge, of } from 'rxjs';
import { concatAll, scan } from 'rxjs/operators';
import contractVersions from '../../versions.json';
import { DEFAULT_CONTRACT_VERSION } from '../helpers/config.helper';
import { defaultUploadOptions } from '../helpers/config.helper';
import { ipfsUpload, prepareImageForLSP3 } from '../helpers/uploader.helper';
import { getUniversalProfileBaseContractAddresses$, universalProfileBaseContractsDeployment$, } from '../services/base-contract.service';
import { keyManagerDeployment$ } from '../services/key-manager.service';
import { accountDeployment$, getTransferOwnershipTransaction$, lsp3ProfileUpload$, setDataTransaction$, } from './../services/lsp3-account.service';
import { universalReceiverDelegateDeployment$ } from './../services/universal-receiver.service';
/**
 * TODO: docs
 */
export class LSP3UniversalProfile {
    constructor(options) {
        this.options = options;
        this.signer = new NonceManager(options.signer);
    }
    /**
     * TODO: docs
     */
    deployReactive(profileDeploymentOptions, contractDeploymentOptions) {
        var _a, _b, _c, _d, _e;
        // -1 > Run IPFS upload process in parallel with contract deployment
        const lsp3Profile$ = lsp3ProfileUpload$(profileDeploymentOptions.lsp3Profile);
        // 0 > Check for existing base contracts and deploy
        const defaultUPBaseContractAddress = (_b = (_a = contractVersions[this.options.chainId]) === null || _a === void 0 ? void 0 : _a.baseContracts) === null || _b === void 0 ? void 0 : _b.ERC725Account[DEFAULT_CONTRACT_VERSION];
        const defaultUniversalReceiverBaseContractAddress = (_d = (_c = contractVersions[this.options.chainId]) === null || _c === void 0 ? void 0 : _c.baseContracts) === null || _d === void 0 ? void 0 : _d.UniversalReceiverDelegate[DEFAULT_CONTRACT_VERSION];
        const defaultBaseContractByteCode$ = forkJoin([
            this.getDeployedByteCode(defaultUPBaseContractAddress !== null && defaultUPBaseContractAddress !== void 0 ? defaultUPBaseContractAddress : '0x0000000000000000000000000000000000000000'),
            this.getDeployedByteCode(defaultUniversalReceiverBaseContractAddress !== null && defaultUniversalReceiverBaseContractAddress !== void 0 ? defaultUniversalReceiverBaseContractAddress : '0x0000000000000000000000000000000000000000'),
        ]);
        const baseContractAddresses$ = getUniversalProfileBaseContractAddresses$(defaultUPBaseContractAddress, defaultUniversalReceiverBaseContractAddress, defaultBaseContractByteCode$, this.signer, contractDeploymentOptions);
        const controllerAddresses = profileDeploymentOptions.controllingAccounts.map((controller) => {
            return typeof controller === 'string' ? controller : controller.address;
        });
        // 1 > deploys ERC725Account
        const account$ = accountDeployment$(this.signer, controllerAddresses, baseContractAddresses$);
        // 2 > deploys KeyManager
        const keyManager$ = keyManagerDeployment$(this.signer, account$, (_e = contractDeploymentOptions === null || contractDeploymentOptions === void 0 ? void 0 : contractDeploymentOptions.libAddresses) === null || _e === void 0 ? void 0 : _e.keyManagerInit);
        // 3 > deploys UniversalReceiverDelegate
        const universalReceiver$ = universalReceiverDelegateDeployment$(this.signer, account$, baseContractAddresses$);
        // 4 > set permissions, profile and universal
        const setData$ = setDataTransaction$(this.signer, account$, universalReceiver$, profileDeploymentOptions.controllingAccounts, lsp3Profile$);
        // 5 > transfersOwnership to KeyManager
        const transferOwnership$ = getTransferOwnershipTransaction$(this.signer, account$, keyManager$);
        return concat([
            account$,
            merge(universalReceiver$, keyManager$),
            setData$,
            transferOwnership$,
        ]).pipe(concatAll());
    }
    /**
     * TODO: docs
     */
    deploy(profileDeploymentOptions, contractDeploymentOptions) {
        const deployments$ = this.deployReactive(profileDeploymentOptions, contractDeploymentOptions).pipe(scan((accumulator, deploymentEvent) => {
            if (deploymentEvent.receipt && deploymentEvent.receipt.contractAddress) {
                accumulator[deploymentEvent.contractName] = {
                    address: deploymentEvent.receipt.contractAddress,
                    receipt: deploymentEvent.receipt,
                };
            }
            return accumulator;
        }, {}));
        return lastValueFrom(deployments$);
    }
    getDeployedByteCode(contractAddress) {
        return this.options.provider.getCode(contractAddress);
    }
    deployBaseContracts() {
        const baseContractsToDeploy$ = of([true, true]);
        const baseContracts$ = universalProfileBaseContractsDeployment$(this.signer, baseContractsToDeploy$);
        const deployments$ = baseContracts$.pipe(scan((accumulator, deploymentEvent) => {
            if (deploymentEvent.receipt && deploymentEvent.receipt.contractAddress) {
                accumulator[deploymentEvent.contractName] = {
                    address: deploymentEvent.receipt.contractAddress,
                    receipt: deploymentEvent.receipt,
                };
            }
            return accumulator;
        }, {}));
        return lastValueFrom(deployments$);
    }
    /**
     * Pre-deploys the latest Version of the LSP3UniversalProfile smart-contracts.
     *
     * @param {'string'} [version] Instead of deploying the latest Version you can also deploy a specific
     *  version of the smart-contracts. A list of all available version is available here.
     */
    preDeployContracts(version) {
        return __awaiter(this, void 0, void 0, function* () {
            console.log(version);
        });
    }
    /**
     * Uploads the LSP3Profile to the desired endpoint. This can be an `https` URL either pointing to
     * a public, centralized storage endpoint or an IPFS Node / Cluster
     *
     * @param {ProfileDataBeforeUpload} profileData
     * @return {*}  {(Promise<AddResult | string>)}
     * @memberof LSP3UniversalProfile
     */
    static uploadProfileData(profileData, uploadOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            uploadOptions = uploadOptions || defaultUploadOptions;
            const [profileImage, backgroundImage] = yield Promise.all([
                prepareImageForLSP3(uploadOptions, profileData.profileImage),
                prepareImageForLSP3(uploadOptions, profileData.backgroundImage),
            ]);
            const profile = {
                LSP3Profile: Object.assign(Object.assign({}, profileData), { profileImage,
                    backgroundImage }),
            };
            // TODO: allow simple http upload too
            const uploadResponse = yield ipfsUpload(JSON.stringify(profile), uploadOptions.ipfsClientOptions);
            return {
                profile,
                url: uploadResponse.cid ? 'ipfs://' + uploadResponse.cid : 'https upload TBD',
            };
        });
    }
}
