// @lukso/lsp-factory.js/build/module/src/lib/classes/proxy-deployer.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { NonceManager } from '@ethersproject/experimental';
import { LSP1UniversalReceiverDelegateInit__factory, UniversalProfileInit__factory, } from '../../';
import { getProxyByteCode } from '../helpers/deployment.helper';
export class ProxyDeployer {
    constructor(signer) {
        this.signer = new NonceManager(signer);
    }
    deployBaseContracts() {
        return __awaiter(this, void 0, void 0, function* () {
            const universalProfile = yield new UniversalProfileInit__factory(this.signer).deploy();
            const universalReceiverDelegate = yield new LSP1UniversalReceiverDelegateInit__factory(this.signer).deploy();
            this.baseContracts = {
                universalProfile,
                universalReceiverDelegate,
            };
            return this.baseContracts;
        });
    }
    deployProxyContract(contract) {
        return __awaiter(this, void 0, void 0, function* () {
            const proxyTx = yield this.signer.sendTransaction({
                data: getProxyByteCode(contract.address),
            });
            const proxyReceipt = yield proxyTx.wait();
            return contract.attach(proxyReceipt.contractAddress);
        });
    }
}
