// @lukso/lsp-factory.js/build/main/src/lib/classes/digital-asset.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DigitalAsset = void 0;
const experimental_1 = require("@ethersproject/experimental");
const rxjs_1 = require("rxjs");
const versions_json_1 = __importDefault(require("../../versions.json"));
const config_helper_1 = require("../helpers/config.helper");
const base_contract_service_1 = require("../services/base-contract.service");
const digital_asset_service_1 = require("../services/digital-asset.service");
class DigitalAsset {
    constructor(options) {
        this.options = options;
        this.signer = new experimental_1.NonceManager(options.signer);
    }
    // LSP7
    deployLSP7DigitalAssetReactive(digitalAssetDeploymentOptions, contractDeploymentOptions) {
        var _a, _b, _c;
        const digitalAsset$ = (0, digital_asset_service_1.lsp7DigitalAssetDeployment$)(this.signer, digitalAssetDeploymentOptions, (_a = contractDeploymentOptions === null || contractDeploymentOptions === void 0 ? void 0 : contractDeploymentOptions.libAddress) !== null && _a !== void 0 ? _a : (_c = (_b = versions_json_1.default[this.options.chainId]) === null || _b === void 0 ? void 0 : _b.baseContracts) === null || _c === void 0 ? void 0 : _c.LSP7DigitalAsset[config_helper_1.DEFAULT_CONTRACT_VERSION]);
        return digitalAsset$;
    }
    deployLSP7DigitalAsset(digitalAssetDeploymentOptions, contractDeploymentOptions) {
        const deployments$ = this.deployLSP7DigitalAssetReactive(digitalAssetDeploymentOptions, contractDeploymentOptions).pipe((0, rxjs_1.scan)((accumulator, deploymentEvent) => {
            if (deploymentEvent.receipt && deploymentEvent.receipt.contractAddress) {
                accumulator[deploymentEvent.contractName] = {
                    address: deploymentEvent.receipt.contractAddress,
                    receipt: deploymentEvent.receipt,
                };
            }
            return accumulator;
        }, {}));
        return (0, rxjs_1.lastValueFrom)(deployments$);
    }
    // LSP8
    deployLSP8IdentifiableDigitalAssetReactive(digitalAssetDeploymentOptions, contractDeploymentOptions) {
        var _a, _b, _c;
        const digitalAsset$ = (0, digital_asset_service_1.lsp8IdentifiableDigitalAssetDeployment$)(this.signer, digitalAssetDeploymentOptions, (_a = contractDeploymentOptions === null || contractDeploymentOptions === void 0 ? void 0 : contractDeploymentOptions.libAddress) !== null && _a !== void 0 ? _a : (_c = (_b = versions_json_1.default[this.options.chainId]) === null || _b === void 0 ? void 0 : _b.baseContracts) === null || _c === void 0 ? void 0 : _c.LSP8IdentifiableDigitalAsset[config_helper_1.DEFAULT_CONTRACT_VERSION]);
        return digitalAsset$;
    }
    deployLSP8IdentifiableDigitalAsset(digitalAssetDeploymentOptions, ContractDeploymentOptions) {
        const deployments$ = this.deployLSP8IdentifiableDigitalAssetReactive(digitalAssetDeploymentOptions, ContractDeploymentOptions).pipe((0, rxjs_1.scan)((accumulator, deploymentEvent) => {
            if (deploymentEvent.receipt && deploymentEvent.receipt.contractAddress) {
                accumulator[deploymentEvent.contractName] = {
                    address: deploymentEvent.receipt.contractAddress,
                    receipt: deploymentEvent.receipt,
                };
            }
            return accumulator;
        }, {}));
        return (0, rxjs_1.lastValueFrom)(deployments$);
    }
    deployBaseContracts() {
        const baseContractsToDeploy$ = (0, rxjs_1.of)([true, true]);
        const baseContracts$ = (0, base_contract_service_1.digitalAssetBaseContractsDeployment$)(this.signer, baseContractsToDeploy$);
        const deployments$ = baseContracts$.pipe((0, rxjs_1.scan)((accumulator, deploymentEvent) => {
            if (deploymentEvent.receipt && deploymentEvent.receipt.contractAddress) {
                accumulator[deploymentEvent.contractName] = {
                    address: deploymentEvent.receipt.contractAddress,
                    receipt: deploymentEvent.receipt,
                };
            }
            return accumulator;
        }, {}));
        return (0, rxjs_1.lastValueFrom)(deployments$);
    }
}
exports.DigitalAsset = DigitalAsset;
