// @lukso/lsp-factory.js/build/main/src/lib/classes/lsp3-universal-profile.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LSP3UniversalProfile = void 0;
const experimental_1 = require("@ethersproject/experimental");
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
const versions_json_1 = __importDefault(require("../../versions.json"));
const config_helper_1 = require("../helpers/config.helper");
const config_helper_2 = require("../helpers/config.helper");
const uploader_helper_1 = require("../helpers/uploader.helper");
const base_contract_service_1 = require("../services/base-contract.service");
const key_manager_service_1 = require("../services/key-manager.service");
const lsp3_account_service_1 = require("./../services/lsp3-account.service");
const universal_receiver_service_1 = require("./../services/universal-receiver.service");
/**
 * TODO: docs
 */
class LSP3UniversalProfile {
    constructor(options) {
        this.options = options;
        this.signer = new experimental_1.NonceManager(options.signer);
    }
    /**
     * TODO: docs
     */
    deployReactive(profileDeploymentOptions, contractDeploymentOptions) {
        var _a, _b, _c, _d, _e;
        // -1 > Run IPFS upload process in parallel with contract deployment
        const lsp3Profile$ = (0, lsp3_account_service_1.lsp3ProfileUpload$)(profileDeploymentOptions.lsp3Profile);
        // 0 > Check for existing base contracts and deploy
        const defaultUPBaseContractAddress = (_b = (_a = versions_json_1.default[this.options.chainId]) === null || _a === void 0 ? void 0 : _a.baseContracts) === null || _b === void 0 ? void 0 : _b.ERC725Account[config_helper_1.DEFAULT_CONTRACT_VERSION];
        const defaultUniversalReceiverBaseContractAddress = (_d = (_c = versions_json_1.default[this.options.chainId]) === null || _c === void 0 ? void 0 : _c.baseContracts) === null || _d === void 0 ? void 0 : _d.UniversalReceiverDelegate[config_helper_1.DEFAULT_CONTRACT_VERSION];
        const defaultBaseContractByteCode$ = (0, rxjs_1.forkJoin)([
            this.getDeployedByteCode(defaultUPBaseContractAddress !== null && defaultUPBaseContractAddress !== void 0 ? defaultUPBaseContractAddress : '0x0000000000000000000000000000000000000000'),
            this.getDeployedByteCode(defaultUniversalReceiverBaseContractAddress !== null && defaultUniversalReceiverBaseContractAddress !== void 0 ? defaultUniversalReceiverBaseContractAddress : '0x0000000000000000000000000000000000000000'),
        ]);
        const baseContractAddresses$ = (0, base_contract_service_1.getUniversalProfileBaseContractAddresses$)(defaultUPBaseContractAddress, defaultUniversalReceiverBaseContractAddress, defaultBaseContractByteCode$, this.signer, contractDeploymentOptions);
        const controllerAddresses = profileDeploymentOptions.controllingAccounts.map((controller) => {
            return typeof controller === 'string' ? controller : controller.address;
        });
        // 1 > deploys ERC725Account
        const account$ = (0, lsp3_account_service_1.accountDeployment$)(this.signer, controllerAddresses, baseContractAddresses$);
        // 2 > deploys KeyManager
        const keyManager$ = (0, key_manager_service_1.keyManagerDeployment$)(this.signer, account$, (_e = contractDeploymentOptions === null || contractDeploymentOptions === void 0 ? void 0 : contractDeploymentOptions.libAddresses) === null || _e === void 0 ? void 0 : _e.keyManagerInit);
        // 3 > deploys UniversalReceiverDelegate
        const universalReceiver$ = (0, universal_receiver_service_1.universalReceiverDelegateDeployment$)(this.signer, account$, baseContractAddresses$);
        // 4 > set permissions, profile and universal
        const setData$ = (0, lsp3_account_service_1.setDataTransaction$)(this.signer, account$, universalReceiver$, profileDeploymentOptions.controllingAccounts, lsp3Profile$);
        // 5 > transfersOwnership to KeyManager
        const transferOwnership$ = (0, lsp3_account_service_1.getTransferOwnershipTransaction$)(this.signer, account$, keyManager$);
        return (0, rxjs_1.concat)([
            account$,
            (0, rxjs_1.merge)(universalReceiver$, keyManager$),
            setData$,
            transferOwnership$,
        ]).pipe((0, operators_1.concatAll)());
    }
    /**
     * TODO: docs
     */
    deploy(profileDeploymentOptions, contractDeploymentOptions) {
        const deployments$ = this.deployReactive(profileDeploymentOptions, contractDeploymentOptions).pipe((0, operators_1.scan)((accumulator, deploymentEvent) => {
            if (deploymentEvent.receipt && deploymentEvent.receipt.contractAddress) {
                accumulator[deploymentEvent.contractName] = {
                    address: deploymentEvent.receipt.contractAddress,
                    receipt: deploymentEvent.receipt,
                };
            }
            return accumulator;
        }, {}));
        return (0, rxjs_1.lastValueFrom)(deployments$);
    }
    getDeployedByteCode(contractAddress) {
        return this.options.provider.getCode(contractAddress);
    }
    deployBaseContracts() {
        const baseContractsToDeploy$ = (0, rxjs_1.of)([true, true]);
        const baseContracts$ = (0, base_contract_service_1.universalProfileBaseContractsDeployment$)(this.signer, baseContractsToDeploy$);
        const deployments$ = baseContracts$.pipe((0, operators_1.scan)((accumulator, deploymentEvent) => {
            if (deploymentEvent.receipt && deploymentEvent.receipt.contractAddress) {
                accumulator[deploymentEvent.contractName] = {
                    address: deploymentEvent.receipt.contractAddress,
                    receipt: deploymentEvent.receipt,
                };
            }
            return accumulator;
        }, {}));
        return (0, rxjs_1.lastValueFrom)(deployments$);
    }
    /**
     * Pre-deploys the latest Version of the LSP3UniversalProfile smart-contracts.
     *
     * @param {'string'} [version] Instead of deploying the latest Version you can also deploy a specific
     *  version of the smart-contracts. A list of all available version is available here.
     */
    async preDeployContracts(version) {
        console.log(version);
    }
    /**
     * Uploads the LSP3Profile to the desired endpoint. This can be an `https` URL either pointing to
     * a public, centralized storage endpoint or an IPFS Node / Cluster
     *
     * @param {ProfileDataBeforeUpload} profileData
     * @return {*}  {(Promise<AddResult | string>)}
     * @memberof LSP3UniversalProfile
     */
    static async uploadProfileData(profileData, uploadOptions) {
        uploadOptions = uploadOptions || config_helper_2.defaultUploadOptions;
        const [profileImage, backgroundImage] = await Promise.all([
            (0, uploader_helper_1.prepareImageForLSP3)(uploadOptions, profileData.profileImage),
            (0, uploader_helper_1.prepareImageForLSP3)(uploadOptions, profileData.backgroundImage),
        ]);
        const profile = {
            LSP3Profile: Object.assign(Object.assign({}, profileData), { profileImage,
                backgroundImage }),
        };
        // TODO: allow simple http upload too
        const uploadResponse = await (0, uploader_helper_1.ipfsUpload)(JSON.stringify(profile), uploadOptions.ipfsClientOptions);
        return {
            profile,
            url: uploadResponse.cid ? 'ipfs://' + uploadResponse.cid : 'https upload TBD',
        };
    }
}
exports.LSP3UniversalProfile = LSP3UniversalProfile;
