// @lukso/lsp-factory.js/build/main/src/lib/classes/proxy-deployer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProxyDeployer = void 0;
const experimental_1 = require("@ethersproject/experimental");
const __1 = require("../../");
const deployment_helper_1 = require("../helpers/deployment.helper");
class ProxyDeployer {
    constructor(signer) {
        this.signer = new experimental_1.NonceManager(signer);
    }
    async deployBaseContracts() {
        const universalProfile = await new __1.UniversalProfileInit__factory(this.signer).deploy();
        const universalReceiverDelegate = await new __1.LSP1UniversalReceiverDelegateInit__factory(this.signer).deploy();
        this.baseContracts = {
            universalProfile,
            universalReceiverDelegate,
        };
        return this.baseContracts;
    }
    async deployProxyContract(contract) {
        const proxyTx = await this.signer.sendTransaction({
            data: (0, deployment_helper_1.getProxyByteCode)(contract.address),
        });
        const proxyReceipt = await proxyTx.wait();
        return contract.attach(proxyReceipt.contractAddress);
    }
}
exports.ProxyDeployer = ProxyDeployer;
